### ISSUE TITLE
Performance of Stability cannot match the paper

### ISSUE THREAD

**Issue by kalifadan** [RL4M/KeAP#3]
Hey!
I'm trying to reproduced your results of the stability task with the KeAP model.
I trained the model on 4 GPUs with 300K steps total using your `run_pretrain.sh` scripts and I succeeded to reproduced the remote_homology task and almost all results in the contact task. However in the stability task the results are:
`metrics {'test_loss': 0.1344011127948761, 'test_spearmanr': 0.7343695245732512, 'test_runtime': 23.5466, 'test_samples_per_second': 545.768, 'test_steps_per_second': 34.145}`.
Do you have an idea what can I do?
Thank you!!!

**Comment by yunxiangfu2001** [comment]
Hi,
I have updated the script. The stability task should be run using the run_stability.py file, which has EarlyStoppingCallback.

**Comment by kalifadan** [comment]
Thank you!!
Also, in the contact prediction, I saw that you split the results by the seq. len.
How it reflected in the code?
How can I run it in the same way?

**Comment by yunxiangfu2001** [comment]
We followed [OntoProtein](https://github.com/zjunlp/OntoProtein) for the contact task. You need to first change the [compute_precision_at_l](https://github.com/RL4M/KeAP/blob/main/replace_code/tape/modeling_utils.py#L875C1-L875C1) functions at [tape/modeling_utils.py](https://github.com/RL4M/KeAP/blob/main/replace_code/tape/modeling_utils.py) and edit the corresponding sequence lengths. Finally, use tape/modeling_utils.py to replace the modeling_utils.py in the TAPE library.

**Comment by kalifadan** [comment]
Great :)
So for reproduce your results on the contact prediction, specifically seq length between 6 to 12 I need to change the function `compute_precision_at_l5` for example to be: `valid_mask &= ((y_ind - x_ind) >= 6 && (y_ind - x_ind) <= 12).unsqueeze(0)`. Is it what you did also?

**Comment by kalifadan** [comment]
> Hi, I have updated the script. The stability task should be run using the run_stability.py file, which has EarlyStoppingCallback.

The new results is still not as the paper shown:
`metrics {'test_loss': 0.10668420046567917, 'test_spearmanr': 0.7420031873337786, 'test_runtime': 22.9074, 'test_samples_per_second': 560.997, 'test_steps_per_second': 35.098}`.
What also can I check?
Thank you!

**Comment by yunxiangfu2001** [comment]
Hi. Sorry for the late reply.

> Great :) So for reproduce your results on the contact prediction, specifically seq length between 6 to 12 I need to change the function `compute_precision_at_l5` for example to be: `valid_mask &= ((y_ind - x_ind) >= 6 && (y_ind - x_ind) <= 12).unsqueeze(0)`. Is it what you did also?

Thanks for your question. 
For the contact task, we followed procedures from OntoProtein strictly with the details provided by the authors. Specifically, the code for computing 6<= seq <12 contact precision is `valid_mask &= ((y_ind - x_ind) >= 6).unsqueeze(0)`. Likewise, `valid_mask &= ((y_ind - x_ind) >= 12).unsqueeze(0)` corresponds to 12<= seq <24. In the paper, we write the results of `valid_mask &= ((y_ind - x_ind) >= 6).unsqueeze(0)` to `6<= seq <12` to be consistent with OntoProtein.


> The new results is still not as the paper shown: `metrics {'test_loss': 0.10668420046567917, 'test_spearmanr': 0.7420031873337786, 'test_runtime': 22.9074, 'test_samples_per_second': 560.997, 'test_steps_per_second': 35.098}`. What also can I check? Thank you!

We have re-run the stability task and found consistent results `{'test_loss': 0.06810372322797775, 'test_spearmanr': 0.8153707581754016, 'test_runtime': 51.1336, 'test_samples_per_second': 251.322, 'test_steps_per_second': 15.724}`. The fine-tuned model can be found at this [link](https://drive.google.com/file/d/1Wh2jouDTK0yznbhGa38I6cCmDIvpsNdu/view?usp=sharing). I'm not sure what is the exact problem. Nonetheless, EarlyStoppingCallback is particularly helpful for stability task and you can try decreasing the evaluation step size to leverage it more effectively. 

I hope this addresses your concerns :)

**Comment by kalifadan** [comment]
Hey,
I'm still having trouble to reproduce the stability results of KeAP.

I used your pretrained model as you publish in the repo ([pre-trained model](https://drive.google.com/file/d/1CZFV8DA4l9F74ias1fR8mHdf1grrjsNq/view?usp=sharing)) and try to fine-tune it on the stability task.
The results are the following:
`metrics {'test_loss': 0.1169305145740509, 'test_spearmanr': 0.7805298494821018, 'test_runtime': 38.4654, 'test_samples_per_second': 334.092, 'test_steps_per_second': 20.902}` which is different from the `0.82` result you published.

I'm running the script in this way (using the `scripts/run_stability.sh`):
`nohup sh run_stability.sh \
      --model "KeAP" \
      --tokenizer "Rostlab/prot_bert" \
      --evaluation_strategy "steps" \
      --metric_for_best_model "eval_spearmanr" \
      --output_file stability-results \
      --task_name stability \
      --do_train True \
      --epoch 5 \
      --mean_output False \
      --optimizer AdamW \
      --per_device_batch_size 2 \
      --gradient_accumulation_steps 32 \
      --eval_step 500 \
      --eval_batchsize 16 \
      --warmup_ratio 0.08 \
      --learning_rate 1e-5 \
      --seed 3 \
      --frozen_bert False > output/stability/KeAP-results.out 2>&1`

If you have any idea I would like to know :)
Thank you!!

**Comment by yunxiangfu2001** [comment]
Maybe try to set eval_step to a small number (i.e, 20)?

**Comment by kalifadan** [comment]
Hey!
I have tried it but it didn't work :(
The results are:
`manr': 0.7805301540769599, 'test_runtime': 39.3397, 'test_samples_per_second': 326.667, 'test_steps_per_second': 16.345}`.
What can I check now?
If you run it, you success to reproduce the results?
You told me before that you have updated the script of the stability check (3 weeks ago) - maybe have you change something there comapre to the original implementation accidentally? (I didn't found somthing).
Thank you :) It will help me a lot if we will find the problem.

**Comment by kalifadan** [comment]
Hey, do you have any updates?
Thank you :)

**Comment by yunxiangfu2001** [comment]
> We have re-run the stability task and found consistent results `{'test_loss': 0.06810372322797775, 'test_spearmanr': 0.8153707581754016, 'test_runtime': 51.1336, 'test_samples_per_second': 251.322, 'test_steps_per_second': 15.724}`. The fine-tuned model can be found at this [link](https://drive.google.com/file/d/1Wh2jouDTK0yznbhGa38I6cCmDIvpsNdu/view?usp=sharing). I'm not sure what is the exact problem. Nonetheless, EarlyStoppingCallback is particularly helpful for stability tasks and you can try decreasing the evaluation step size to leverage it more effectively.

Hi, we checked the scripts and did not find any difference and successfully reproduce the results with the checkpoint here. The results may fluctuate a bit, but shouldn't be as low as 0.78. 

Note: we found that setting eval_step to 20 leads to a small increase in test_spearmanr.
